draft/high-level-architecture.rst

```diff
@@ -1,106 +1,162 @@
+The rest are replaced by new ones which provide more functionality and flexibility for our users.
+
+In the rest of this specification we will describe the architectural building blocks of Celery.
+
+Message Broker
+--------------
+
+In Celery 4.x each :term:`Celery Master` connected to only one :term:`Message Broker` cluster.
+
+This is no longer the case.
+Celery now allows connecting to multiple :term:`Message Brokers <Message Broker>`
+even if they are of clusters that use different implementations of a message broker.
+
+Users can consume messages from a Redis cluster, a RabbitMQ cluster, and an ActiveMQ cluster if they so desire.
+
+This feature is useful when, for example:
+
+- The user migrates from a legacy system that uses other implementation of a :term:`Message Broker`, but the new system uses a more modern one.
+- The user wants to split the load between clusters.
+- There's a security reason to publish some messages to a specific cluster.
+
+On some :term:`Message Broker` implementations the `Controller`_ will assist in managing the cluster.
+
+Data Sources & Sinks
+--------------------
+
+In Celery 4.x we had a :term:`Result Backend` which was used to store task results and coordinate
+the execution of chords.
+
+We extend the :term:`Result Backend` concept further to allow new use cases such:
+
+- :term:`ETL`.
+- :term:`Data Integration`.
+- Reporting.
+- Taking action when data is inserted or updated.
+
+In addition, like we did for the `Message Broker`_, we now allow multiple data sources and sinks
+instead of one cluster of a :term:`Result Backend`.
+
+The responsibility for coordination of the execution of chords has moved to the `Execution Platform`_.
+
+Data Sinks
+~~~~~~~~~~
+
+A data sink is where task results are saved.
+
+A task result may be saved in more than one data sink (e.g. a Kafka Topic and S3).
+
+The `Router`_ is responsible for routing task results to the correct data sink(s) and properly
+serializing them.
+
+Data Sources
+~~~~~~~~~~~~
+
+Controller
+----------
+
+Router
+------
+
+Execution Platform
+------------------
 
 Motivation
 ==========
 
-This section should explain *why* this CEP is needed. The motivation is critical
-for CEPs that want to add substantial new features or materially refactor
-existing ones.  It should clearly explain why the existing solutions are
-inadequate to address the problem that the CEP solves.  CEP submissions without
-sufficient motivation may be rejected outright.
+We want to modernize Celery for the Cloud Native age.
+We need to keep Celery relevant for our users and help them in new ways. Therefore, we must adjust and evolve to meet the unique challenges of the Cloud Native age.
 
-Rationale
-=========
+Also, we want to modernize the code to support Python 3+, which will allow us to remove workarounds, backports, and compatibility shims.
+Refactoring the codebase to support Python 3+ allows us to keep a slimmer, more maintainable codebase.
```

Comment:
3.6/3.8+?

Reply:
I'd rather avoid being specific in this document.
My last attempt was too specific.
We will specify our supported versions in another document that is more detailed about Celery 5.0.

If you think otherwise, let me know.

Reply:
OK cool